src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -344,6 +341,15 @@ public void onError(Session session, Throwable cause) {
 		RelayLogger.throwingException(cause, this);
 	}
 	
+	private static ByteBuffer deepCopyByteBuffer(ByteBuffer original) {
```

Comment:
>deepCopyByteBuffer [](start = 27, length = 18)

Since I'm still trying to wrap my brain around ByteBuffer without reading the docs for every call I wrote this and it all passed.
```
 // Test
 public void cloneByteBufferTest() {
  byte[] array = new byte[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
  ByteBuffer source = ByteBuffer.wrap(array);
  ByteBuffer clone = deepCopyByteBuffer(source);
  assertByteBufferEquals(source, clone);

  source = ByteBuffer.wrap(array, 0, array.length / 2);
  clone = deepCopyByteBuffer(source);
  assertByteBufferEquals(source, clone);

  source = ByteBuffer.wrap(array, 1, array.length - 2);
  clone = deepCopyByteBuffer(source);
  assertByteBufferEquals(source, clone);
 }

  static void assertByteBufferEquals(ByteBuffer expected, ByteBuffer actual) {
  assertEquals(expected.capacity(), actual.capacity());  
  assertEquals(expected.arrayOffset(), actual.arrayOffset());
  assertEquals(expected.hasArray(), actual.hasArray());
  assertEquals(expected.remaining(), actual.remaining());
  assertEquals(expected.limit(), actual.limit());
  assertEquals(expected.array().length, actual.array().length);
  assertNotEquals(expected.array(), actual.array());

  while (expected.hasRemaining() && actual.hasRemaining()) {
   byte expectedValue = expected.get();
   byte actualValue = actual.get();
   assertEquals(expectedValue, actualValue);
  }

  assertEquals(expected.hasRemaining(), actual.hasRemaining());
 }
```

Reply:
Here's a potential way to fix these issues:
```
 private static ByteBuffer deepCopyByteBuffer2(ByteBuffer original) {
  int originalPosition = original.position();
  ByteBuffer clone = ByteBuffer.wrap(new byte[original.capacity()], original.arrayOffset(), original.limit());
  original.rewind();
  clone.put(original);

  original.position(originalPosition);
  clone.position(originalPosition);

  return clone;
  }
```

---
In reply to: [296344691](https://github.com/Azure/azure-relay-java/pull/48#discussion_r296344691) [](ancestors = 296344691)

Reply:
It's possible I went a bit above and beyond what may be functionally needed. Taking a step back I created this PR using ByteBuffers with our public API.

[https://github.com/Azure/azure-relay-java/pull/49](https://github.com/Azure/azure-relay-java/pull/49)

It would be awesome to see if that new test I added works with iteration 1 of deepCopyByteBuffer.

---
In reply to: [296371319](https://github.com/Azure/azure-relay-java/pull/48#discussion_r296371319) [](ancestors = 296371319,296344691)